math/optimized_trig.h

```diff
@@ -0,0 +1,233 @@
+ *                        d = 1/(G^6)*(G/2 - 2 + b*G^2);  // 0.0012440842722145661
+ *
+ *                     Sine is calculated with the same equaiton by recognizing that sin(x) = cos(x+pi/2)
+ *
+ *                     A transformation of z = x/pi is used to map the circle into the field
+ *                     of [-1, 1), which leads itself to easy fixed-point processing.
+ *
+ *
+ *                     NOTE: Only arguments between (-2*pi and +2pi) are valid. Values outside
+ *                           that range will cause integer overflow and the algorithm will silently
+ *                           fail.
+ *                     NOTENOTE: Please read the above note on the valid range.
+ *                     NOTENOTENOTE: Srsly, only values in [pi, -pi) should be used, and never, ever
+ *                                   exceed (-2*pi+1) and (2*pi-1).
+ * @param angle_rad[float] The angle in [radians], in the range of [-pi,pi)
+ * @param sin_out[float] The 6th order approximation of sin(x)
+ * @param cos_out[float] The 6th order approximation of cos(x)
+ */
+__attribute__((always_inline))
+inline static void _SinCos66_FP(const float angle_rad, float *sin_out, float *cos_out) {
+   const uint8_t qN = 28;
+   const uint8_t qX = 16;
+   const uint8_t qQ = 30;
+
+   const float PI_INVERSE = 1.0f/M_PI;  // Forces the compiler to precompute the inverse of PI
+   const int32_t FP_PI_Q = (1<<qQ);  // qQ
+   const int32_t FP_PI_HALVES_Q = FP_PI_Q>>1; // qQ
+   const int32_t FP_PI = (1<<qX);  // qX
+   const int32_t FP_PI_HALVES = FP_PI>>1;  // qX
+
+   const double G=M_PI/2;
+   const double G2=G*G;
+   const double G3=G2*G;
+   const double G4=G2*G2;
+   const double G6=G4*G2;
+
+   // a - x^2*((b - x^2*(c - d*x^2))
+   // Coefficients found by minimizing the RMS error, and ensuring that the sine wave passes perfectly by 0 and 1.
+   const double a=1;
+   const double b = (-45045/G4*(G6+2*G4-156*G2+360)/(G3) - (-655*G-4504))/(952*G2); //# 4.933160419860462,;  // 0.499833652837794
+   const double c = 1/(2*G4)*(G - 6 + 4*b*G2);  // 0.041388888146338884
+   const double d = 1/(G6)*(G/2 - 2 + b*G2);  // 0.0012440842722145661
+
+   // Transform x to q, where q = x/pi
+   // aq - q^2*((bq - q^2*(cq - dq*q^2))
+   const double aq = a;
+   const double bq = b * G2*4;
+   const double cq = c * G4*16;
+   const double dq = d * G6*64;
+
+   // Convert the coefficients to fixed point, and inject a negative sign for certain ones. This negative
+   // sign will be useful when using the multiply & accumulate assembly instruction.
+   const int32_t AQ = aq * (1UL<<qN);  // qN
+   const int32_t BQm = -(bq * (1UL<<qN) + 0.5L);  // qN
+   const int32_t CQ = cq * (1UL<<qN) + 0.5L;  // qN
+   const int32_t DQm = -(dq * (1UL<<((48-(32-qN)-2*qX))) + 0.5L);  // ((16 + qN - 2*qX)))
+
+   // Convert the angle to fixed-point.
+   int32_t angle_FP = (angle_rad*PI_INVERSE)  * (1<<qQ); //  Do not round, as it is important not to reach +2^15
```

Comment:
Low order bits are discarded later. It shall be possible to allocate some bits for range outsize -2pi..2pi and normalize argument using bitmasking/simple integer math. 
It will also allow rounding here, increasing accuracy a bit (but later shift must be considered - rounding before 1<<16 division has almost no effect)

Reply:
@ledvinap, that's an excellent observation. It can save some additional cycles because there's no floating-point division to scale back to radians. Plus, the quantization error as cos(x)--> 0 completely disappears.

Reply:
The alternative is to just do
```
         // Ensure angle is in [-pi,pi)
         while (angle_rad < -M_PI) { angle_rad += 2.0 * M_PI; }
         while (angle_rad >  M_PI) { angle_rad -= 2.0 * M_PI; }

         if ((is2ndOr4thQuadrant && !is3rdOr4thQuadrant) || // the 2nd quadrant
             (!is2ndOr4thQuadrant && is3rdOr4thQuadrant)) { // the 3rd quadrant
            output = -M_PI + angle_rad;
         } else {
            output = angle_rad;
         }
```

While that has those while loops, they are (1) unlikely to be looped more than once because the function shouldn't be -passed values in the wrong range-- and if it is then the fixed-point math generally breaks so we no longer care about efficiency-- and (2) it helps even out the processor requirements for both estimation branches (esp. since this branch no longer has the `*1/1<<qN` factor.

I'm inclined to go with this option, as it is much more readable. The downsides only occur when we have extra CPU budget.